A dataquest mission is covered here:
* https://www.dataquest.io/mission/63/modifying-data

In [1]:
import pyspark
sc = pyspark.SparkContext('local[*]')

In [2]:
# do something to prove it works
rdd = sc.parallelize(range(1000))
###rdd.takeSample(False, 5)

In [3]:
### http://www.cdotson.com/2014/06/generating-json-documents-from-sqlite-databases-in-python/
#import sqlite3
#def dict_factory(cursor, row):
#    d = {}
#    for idx, col in enumerate(cursor.description):
#        d[col[0]] = row[idx]
#    return d
# 
#connection = sqlite3.connect("factbook.db")
#connection.row_factory = dict_factory
# 
#cursor = connection.cursor()
# 
#cursor.execute("select * from facts")
# 
# fetch all or one we'll go for all.
# 
#results = cursor.fetchall()
# 
###print(results)
#import sys
#with open('factbook.json', 'w') as f:
#    sys.stdout = f
#    print (results)
#    
#connection.close()

In [4]:
# Import SQLContext
from pyspark.sql import SQLContext

# Pass in the SparkContext object `sc`
sqlCtx = SQLContext(sc)

# Read JSON data into DataFrame object `df`
df00 = sqlCtx.read.json("factbook.json")
f00 = df00.head(5)
for r00 in f00:
    print(r00)
df = df00.drop('_corrupt_record')

Row(_corrupt_record=None, area=652230, area_land=652230, area_water=0, birth_rate=38.57, code='af', created_at='2015-11-01 13:19:49.461734', death_rate=13.89, id=1, migration_rate=1.51, name='Afghanistan', population=32564342, population_growth=2.32, updated_at='2015-11-01 13:19:49.461734')
Row(_corrupt_record=None, area=28748, area_land=27398, area_water=1350, birth_rate=12.92, code='al', created_at='2015-11-01 13:19:54.431082', death_rate=6.58, id=2, migration_rate=3.3, name='Albania', population=3029278, population_growth=0.3, updated_at='2015-11-01 13:19:54.431082')
Row(_corrupt_record=None, area=2381741, area_land=2381741, area_water=0, birth_rate=23.67, code='ag', created_at='2015-11-01 13:19:59.961286', death_rate=4.31, id=3, migration_rate=0.92, name='Algeria', population=39542166, population_growth=1.84, updated_at='2015-11-01 13:19:59.961286')
Row(_corrupt_record=None, area=468, area_land=468, area_water=0, birth_rate=8.13, code='an', created_at='2015-11-01 13:20:03.659945', 

In [5]:
df.printSchema()

root
 |-- area: long (nullable = true)
 |-- area_land: long (nullable = true)
 |-- area_water: long (nullable = true)
 |-- birth_rate: double (nullable = true)
 |-- code: string (nullable = true)
 |-- created_at: string (nullable = true)
 |-- death_rate: double (nullable = true)
 |-- id: long (nullable = true)
 |-- migration_rate: double (nullable = true)
 |-- name: string (nullable = true)
 |-- population: long (nullable = true)
 |-- population_growth: double (nullable = true)
 |-- updated_at: string (nullable = true)



In [6]:
#preparations before running  SQL queries
df.registerTempTable("facts")
tables=sqlCtx.tableNames()
print(tables)
#query = 'SELECT males, females FROM census2010 WHERE age > 5 AND age < 15'
#sqlCtx.sql(query).show()

['facts']


In [7]:
query = """
SELECT * FROM facts 
WHERE updated_at > "2015-10-30 16:00"
AND updated_at < "2015-11-02 15:00"
"""
sqlCtx.sql(query).show()

+-------+---------+----------+----------+----+--------------------+----------+---+--------------+-------------------+----------+-----------------+--------------------+
|   area|area_land|area_water|birth_rate|code|          created_at|death_rate| id|migration_rate|               name|population|population_growth|          updated_at|
+-------+---------+----------+----------+----+--------------------+----------+---+--------------+-------------------+----------+-----------------+--------------------+
| 652230|   652230|         0|     38.57|  af|2015-11-01 13:19:...|     13.89|  1|          1.51|        Afghanistan|  32564342|             2.32|2015-11-01 13:19:...|
|  28748|    27398|      1350|     12.92|  al|2015-11-01 13:19:...|      6.58|  2|           3.3|            Albania|   3029278|              0.3|2015-11-01 13:19:...|
|2381741|  2381741|         0|     23.67|  ag|2015-11-01 13:19:...|      4.31|  3|          0.92|            Algeria|  39542166|             1.84|2015-11-01 13:

This was executed **locally** on a laptop:

<pre>
sqlite3 ~/Downloads/factbook.db 
SQLite version 3.8.4.1 2014-03-11 15:27:36
Enter ".help" for usage hints.
sqlite> PRAGMA table_info(facts);    
0|id|INTEGER|1||1
1|code|varchar(255)|1||0
2|name|varchar(255)|1||0
3|area|integer|0||0
4|area_land|integer|0||0
5|area_water|integer|0||0
6|population|integer|0||0
7|population_growth|float|0||0
8|birth_rate|float|0||0
9|death_rate|float|0||0
10|migration_rate|float|0||0
11|created_at|datetime|0||0
12|updated_at|datetime|0||0
</pre>

In [13]:
# Write a SQL query that uses the ORDER BY and LIMIT statements to 
# select the entire row that has the highest id value
query1 = """
SELECT id,* FROM facts 
ORDER BY id DESC
LIMIT 20
"""
sqlCtx.sql(query1).show()

+---+-----+---------+----------+----------+----+--------------------+----------+---+--------------+--------------------+----------+-----------------+--------------------+
| id| area|area_land|area_water|birth_rate|code|          created_at|death_rate| id|migration_rate|                name|population|population_growth|          updated_at|
+---+-----+---------+----------+----------+----+--------------------+----------+---+--------------+--------------------+----------+-----------------+--------------------+
|254| 5860|     5640|       220|     22.99|  we|2015-11-01 13:38:...|       3.5|254|           0.0|           West Bank|   2785366|             1.95|2015-11-01 13:38:...|
|251|  360|      360|         0|     31.11|  gz|2015-11-01 13:38:...|      3.04|251|           0.0|          Gaza Strip|   1869055|             2.81|2015-11-01 13:38:...|
|247| 1910|      346|      1564|     10.31|  vq|2015-11-01 13:38:...|      8.54|247|          7.67|      Virgin Islands|    103574|             0

This was executed locally on a laptop:

<pre>
sqlite> INSERT INTO facts
VALUES (262, "dq", "DataquestLand", 60000, 40000, 20000, 500000, 100, 50, 10, 20, "2016-02-25 12:00:00", "2016-02-25 12:00:00");
sqlite> SELECT * FROM facts  ORDER BY id DESC LIMIT 12;   
262|dq|DataquestLand|60000|40000|20000|500000|100.0|50.0|10.0|20.0|2016-02-25 12:00:00|2016-02-25 12:00:00
261|xx|World||||7256490011|1.08|18.6|7.8||2015-11-01 13:39:09.910721|2015-11-01 13:39:09.910721
260|oo|Southern Ocean|||||||||2015-11-01 13:39:05.806048|2015-11-01 13:39:05.806048
259|zn|Pacific Ocean|||||||||2015-11-01 13:39:04.179530|2015-11-01 13:39:04.179530
258|xo|Indian Ocean|||||||||2015-11-01 13:39:02.723888|2015-11-01 13:39:02.723888
257|zh|Atlantic Ocean|||||||||2015-11-01 13:39:01.255361|2015-11-01 13:39:01.255361
256|xq|Arctic Ocean|||||||||2015-11-01 13:38:59.784157|2015-11-01 13:38:59.784157
255|wi|Western Sahara|266000|266000|0|570866|2.82|30.24|8.34||2015-11-01 13:38:58.291161|2015-11-01 13:38:58.291161
254|we|West Bank|5860|5640|220|2785366|1.95|22.99|3.5|0.0|2015-11-01 13:38:55.084643|2015-11-01 13:38:55.084643
253|pg|Spratly Islands|5|5|0||||||2015-11-01 13:38:51.305215|2015-11-01 13:38:51.305215
252|pf|Paracel Islands|7|7|0||||||2015-11-01 13:38:49.622992|2015-11-01 13:38:49.622992
251|gz|Gaza Strip|360|360|0|1869055|2.81|31.11|3.04|0.0|2015-11-01 13:38:48.049460|2015-11-01 13:38:48.049460
</pre>

sqlite> SELECT * FROM facts WHERE area IS NULL;
35|cd|Chad||1259200|24800|11631456|1.89|36.6|14.28|3.45|2015-11-01 13:22:42.673800|2015-11-01 13:22:42.673800
128|ng|Niger||1266700|300|18045729|3.25|45.45|12.42|0.56|2015-11-01 13:30:34.606095|2015-11-01 13:30:34.606095
210|fs|French Southern and Antarctic Lands|||||||||2015-11-01 13:36:43.532683|2015-11-01 13:36:43.532683
212|tb|Saint Barthelemy||||7237|||||2015-11-01 13:36:49.817701|2015-11-01 13:36:49.817701
249|um|United States Pacific Island Wildlife Refuges|||||||||2015-11-01 13:38:42.851318|2015-11-01 13:38:42.851318
250|ay|Antarctica||280000||0|||||2015-11-01 13:38:44.885746|2015-11-01 13:38:44.885746
256|xq|Arctic Ocean|||||||||2015-11-01 13:38:59.784157|2015-11-01 13:38:59.784157
257|zh|Atlantic Ocean|||||||||2015-11-01 13:39:01.255361|2015-11-01 13:39:01.255361
258|xo|Indian Ocean|||||||||2015-11-01 13:39:02.723888|2015-11-01 13:39:02.723888
259|zn|Pacific Ocean|||||||||2015-11-01 13:39:04.179530|2015-11-01 13:39:04.179530
260|oo|Southern Ocean|||||||||2015-11-01 13:39:05.806048|2015-11-01 13:39:05.806048
261|xx|World||||7256490011|1.08|18.6|7.8||2015-11-01 13:39:09.910721|2015-11-01 13:39:09.910721
263|dq|DataquestLand|||20000|500000|100.0|50.0|10.0|20.0|2016-02-25 12:00:00|2016-02-25 12:00:00

<pre>
sqlite> UPDATE facts
SET name="DataquestLand", code="dq"
WHERE name="United States";   ...>    ...> 
sqlite> SELECT * FROM facts WHERE id=186;
186|dq|DataquestLand|9826675|9161966|664709|321368864|0.78|12.49|8.15|3.86|2015-11-01 13:35:14.898271|2015-11-01 13:35:14.898271
</pre>                